### ISSUE TITLE
How to package only needed files with package_files?

### ISSUE THREAD

**Issue by drozel** [conan-io/conan#1495]
I have to package local binaries according to [this manual](http://docs.conan.io/en/latest/howtos/existing_binaries.html#packaging-local-binaries).
Project has following files structure:
build/include
build/obj/libs
If i try to 

`conan package_files myLib/1.0@me/main -s os=Linux -s compiler=gcc -s compiler.version=5.4 -pf build`
conan packs full file structure.

I tried to change conanfile.py:
`def package_info(self):
        self.cpp_info.includedirs = ['include']  # Ordered list of include paths
        self.cpp_info.libdirs = ['obj/lib']  # Directories where libraries can be found`
and it did`t help.
Project who has myLib as dependency still searches libs in package/hash/lib.

How i can to say to conan directly: this file to folder include/, these to folder lib/, and i don`t need all other?

**Comment by memsharded** [comment]
The ``package_files`` command will take the files from the current folder, with its current layout and put them into conan local cache. Your approach to customize ``package_info()`` seems good, so it should be a minor detail that is not here. Could you please put a minimum working example in a repo? Please use the ``test_package`` folder for the "consumer" project, that is very convenient.

If you want to pre-process what you have, extract the artifacts for a new layout. You could also implement the ``package()`` method and run the local ``conan package`` command to create a "clean" folder with the state that you want to later put into conan with ``conan package-files``

**Comment by memsharded** [comment]
Just created an example will upload it asap

**Comment by memsharded** [comment]
Please check this example.

[pkgfiles.zip](https://github.com/conan-io/conan/files/1140380/pkgfiles.zip)

You can run ``python build.py`` in the root folder.

**Comment by drozel** [comment]
@memsharded  thank you for your answer.
I`ve uploaded Botan lib to conan_botan repo in my account.
We will assume that i don`t need to build this library, only create package from ready binaries and headers.
The same i can need in other libs, to package results correctly after build done.
I would like to understand, how conan can pack build results for future use

**Comment by memsharded** [comment]
I am not sure if I understand you. The ``build.py`` script contains conan command that can reproduce the workflow that you intend:
- conan export will put the package recipe into conan local cache
- conan package_files put your built library (in the example is just a header)
- The ``package_info()`` method defines the package layout and info for consumers.
- Consumers, like the one inside the ``test_package`` folder can install and link with the project.

Is the example I uploaded working for you? 
Has the Botan lib  ``conan_botan`` repo the pre-build binaries and headers in it? We need a base example we can reproduce. Also I would like to have the output, the exact messages you get when it doesnt work. What I can tell you from a quick look is that you are missing a ``self.cpp_info.libs = ["botan"]`` (or something related) in ``package_info()`` method

**Comment by memsharded** [comment]
Ok, cool, let me know when you have a look to the example. Cheers!

**Comment by drozel** [comment]
@memsharded ok, James! Thank you for your help. I still have questions.
I've built your example and added it to dependencies of my projects. It has 2 dependencies: your Hello and Botan i have in repo:
`Botan/2.1.0@kinexon/main
Hello/0.1@user/channel`
So, i run `conan install` in my project (generator QMake) and see conanbuildinfo.pri result:
`CONAN_INCLUDEPATH += "/home/vitaly/.conan/data/Botan/2.1.0/user/main/package/81607951755e61cf17149d40bed6aba5b3a079a9/include" \
    "/home/vitaly/.conan/data/Hello/0.1/user/channel/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/headers/include"
CONAN_LIBS += 
CONAN_LIBDIRS += -L"/home/vitaly/.conan/data/Botan/2.1.0/user/main/package/81607951755e61cf17149d40bed6aba5b3a079a9/lib" \
    -L"/home/vitaly/.conan/data/Hello/0.1/user/channel/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/lib"`

So you can see, path to headers is been correctly parsed from your package_info()
However, the path to botan libs is been not parsed: Botan conanfile.py contains:
`self.cpp_info.libdirs = ['obj/lib']`
But you can see, generated .pri says to search libs in /home/vitaly/.conan/data/Botan/2.1.0/user/main/package/81607951755e61cf17149d40bed6aba5b3a079a9/lib dir. This is a cause why I become "Undefined reference" error by compiling.

**Comment by memsharded** [comment]
It doesn't matter the way you package or the actual layout, what you get in the ``conanbuildinfo.pri`` file is the translation of what you declared in the package ``package_info()`` method. If the correct path with ``obj/libs`` is not there, is probably because you forgot to run ``conan export`` to put the latest recipe version into conan local cache. This is why setting up an example project, something similar to the one I sent you can help. With a ``build.py`` script that would make sure that we make no mistakes while typing the correct sequence of commands. Can you please setup an example like that? You don't need to have the actual final binary, you can fake it for testing purposes.

**Comment by drozel** [comment]
@memsharded yes, i did it. 
I did't understand, that each conan package has libinfo (properties with lib paths etc.) and one or more exports (libraries and headers). The first one is being experted once with `conan export`, the second one is being exported for platforms and compilers with `conan package_files`. 
It works now, thank you!

If I may, I will ask one more  question:
if "library" generates not .a (who have to be linked with key -l and -L) but .o (and i need to add it as objects at linkage), how i can say it to Conan?

**Comment by memsharded** [comment]
> It works now, thank you!
Excellent :)

For linking objects from another package, I think you might need to do some extra post-process in the consumer side, because such an object model is not implemented out of the box. BTW, why not a library?

So you have full control on the consumer side. You can add your objects to LIBS, like:

```python
self.cpp_info.libs = ["lib1.o", "lib2.o"]
```
Then in your qmake ``project.pro`` (sorry, I don't know much about qmake), you could just filter that, and create your own OBJECTS, so:
```
OBJECTS += $$CONAN_LIBS
CONAN_LIBS = [ ]  # empty it
```
I don't know if it will make a copy or is a reference, but with some qmake-fu, probably can be easily defined. Also, if the CONAN_LIBS might have both libraries and objects, just do a filter that separate them.

**Comment by drozel** [comment]
@memsharded 

> BTW, why not a library?

because gcc can't link an Botan.a to another mylib.a (i comlipe) so that you don't need link Botan in all projects where you do use "mylib".
If mylib is being linked with .o, resulting mylib.a it has Botan internal.

**Comment by memsharded** [comment]
I think that any compiler, not only gcc will link a static library embedded inside another static library (but it will do in a shared library). But just in case, that is one of the good points to use conan, you do not need to embed the objects/library anymore, because conan will handle the transitive dependencies. So when the final application decides to link with your ``mylib.a`` conan will manage to transitively retrieve and link with ``botan.a``. Unless you want to redistribute your ``mylib.a`` without dependencies to your final users, "hiding" that it has a botan dependency. Is that your use case? 
You are very welcome, love to help.

**Comment by drozel** [comment]
> Is that your use case?
No, i'm trying Conan only for internal use. You are right, i do not need it more because i don't need to set Botan as requirements in an application who uses mylib. Because mylib has Botan in its requirements!

**Comment by cleeland** [comment]
FYI, this is related to issue #1489 that I filed a few days back.

> i'm trying Conan only for internal use.

This is exactly my use-case as well, and I think that this use-case is currently not well represented in conan's design and implementation. It's good to see other users who have the same use-case, though, as that will certainly help ensure that conan evolves for both internal-only as well as the external distributions.

**Comment by memsharded** [comment]
> No, i'm trying Conan only for internal use. You are right, i do not need it more because i don't need to set Botan as requirements 

I mean, of course you don't have to set or declare a dependency to Botan in your application. Your application conanfile will have a ``requires=MyLib/0.1@.....`` only. But conan will transitively manage to retrieve Botan lib and add it to the list of libraries that your application has to link with, transparently for you and your final application. So you don't need to manage MyLib embedding Botan objects anymore. And if you use Botan from other library too, lets say MyLib2, that you want to link into your application, it will work fine. With the objects approach you will get duplicated symbols.

So, from my point of view, this approach only would make sense, if you are actually redistributing MyLib as your product, and you want to distribute exactly one file, and hiding that dependency to Botan. Am I missing something from your use case? Thanks for the feedback!

**Comment by drozel** [comment]
>>Am I missing something from your use case? 
No, Conan can everything what i need at this time.
I will try with further libraries i need. Thank you! It was very helpful.

**Comment by memsharded** [comment]
Hi @drozel 

You might be interested in the feature just released in conan 0.26, that allows ``conan package_files`` to use the ``package()`` method to select files to be packaged. So this basically allows skipping the local ``conan package`` command. Check the docs: http://docs.conan.io/en/latest/reference/commands/package_files.html

So I think this issue can be closed now, but please re-open or submit a new issue for any further question. Thanks!
